research/configs/fractal_config.py

```diff
@@ -104,87 +197,92 @@ def a3b3_data():
         }
     )
 
+def ab_grid(*, param='mup', opt='adam', alignment='full', n_layers=3, l=3, resolution=5, ab_range=0.2):
+    cfg = param_frac(param, opt, alignment, n_layers)
+    a_grid = np.linspace(cfg['al'][l - 1] - ab_range, cfg['al'][l - 1] + ab_range, num=resolution).tolist()
+    b_grid = np.linspace(cfg['bl'][l - 1] - ab_range, cfg['bl'][l - 1] + ab_range, num=resolution).tolist()
 
-def mup_a3b3_grid():
-    resolution = 16
-    a3_grid = np.linspace(-0.5, 1.5, num=resolution).tolist()  # Adjust the range as needed
-    b3_grid = np.linspace(-0.5, 1.5, num=resolution).tolist()  # Adjust the range as needed
+    for a_id, a in enumerate(a_grid):
+        for b_id, b in enumerate(b_grid):
+            exp_id = a_id * len(b_grid) + b_id
 
-    for a3_id, a3 in enumerate(a3_grid):
-        for b3_id, b3 in enumerate(b3_grid):
-            exp_id = a3_id * len(b3_grid) + b3_id
-
-            def mup_w_diff_a3b3():
-                cfg = maximal_update_parametrization()
-                cfg['al'][-1] = a3
-                cfg['bl'][-1] = b3
+            def diff_ab():
+                cfg = param_frac(param, opt, alignment, n_layers)
+                cfg['al'][l - 1] = a
+                cfg['bl'][l - 1] = b
                 return cfg
 
-            param_args = (training_frac, mlp2h, sgd_frac, mup_w_diff_a3b3, a3b3_data)
+            param_args = (training_frac, mlp2h, opt_frac(opt), diff_ab, ab_data)
 
-            run_name = f"mup_a3_{a3:.14f}_b3_{b3:.14f}"
+            run_name = f"{param}_{opt}_{alignment}_a{l}_{a:.2f}_b{l}_{b:.2f}"
             yield exp_id, run_name, param_args
 
-def mup_a3b3_eps_grid():
-    sp_resolution = 8
-    t_resolution = 11
-
-    a3_grid = np.linspace(-0.5, 1.5, num=sp_resolution).tolist()
-    b3_grid = np.linspace(-0.5, 1.5, num=sp_resolution).tolist()
+def ab_eps_grid(*, param='mup', opt='adam', alignment='full', n_layers=3, l=3, resolution=5, t_resolution=11, ab_range=0.2):
```

Comment:
what is the `*` for?

Reply:
To force arguments to be passed via kwargs instead of args, but we can remove this - it was just a hint for me while coding to avoid making a mistake